packages/voice/src/app/agent/chat.tsx

```diff
@@ -477,3 +510,175 @@ export class ChatManager {
     this.changeState(ChatManagerState.LISTENING);
   }
 }
+
+export class StreamAnalyzer {
+  source: MediaStreamAudioSourceNode;
+  analyzer: AnalyserNode;
+  constructor(context: AudioContext, stream: MediaStream) {
+    this.source = context.createMediaStreamSource(stream);
+    this.analyzer = context.createAnalyser();
+    this.source.connect(this.analyzer);
+  }
+  stop() {
+    this.source.disconnect();
+  }
+}
+
+/**
+ * Manages a single chat with a LLM, including speculative execution.
+ * All RPCs are performed remotely, and audio is streamed to/from the server via WebRTC.
+ */
+export class WebRtcChatManager implements ChatManager {
+  private params: ChatManagerInit;
+  private audioContext = new AudioContext();
+  private audioElement = new Audio();
+  private textEncoder = new TextEncoder();
+  private textDecoder = new TextDecoder();
+  private _state = ChatManagerState.IDLE;
+  private socket?: WebSocket;
+  private room?: Room;
+  private localAudioTrack?: LocalAudioTrack;
+  private inAnalyzer?: StreamAnalyzer;
+  private outAnalyzer?: StreamAnalyzer;
+  private pinger?: NodeJS.Timer;
+  onStateChange?: (state: ChatManagerState) => void;
+  onInputChange?: (text: string, final: boolean, latency?: number) => void;
+  onOutputChange?: (text: string, final: boolean, latency: number) => void;
+  onAudioGenerate?: (latency: number) => void;
+  onAudioStart?: (latency: number) => void;
+  onAudioEnd?: () => void;
+  onError?: () => void;
+
+  constructor(params: ChatManagerInit) {
+    this.params = params;
+    this.audioElement = new Audio();
+    this.warmup();
+  }
+  get state() {
+    return this._state;
+  }
+  get inputAnalyzer() {
+    return this.inAnalyzer?.analyzer;
+  }
+  get outputAnalyzer() {
+    return this.outAnalyzer?.analyzer;
+  }
+  warmup() {
+    const url = 'ws://localhost:8080/livekit'; //`${window.location.protocol === 'https:' ? 'wss:' : 'ws:'}//${window.location.host}/livekit')`;
```

Comment:
is this what you want?

Reply:
Until we have a prod deploy, yes, I think

Reply:
This is what I see when trying to run the preview:

<img width="594" alt="Screen Shot 2023-11-13 at 5 37 08 PM" src="https://github.com/fixie-ai/ai-jsx/assets/130235/6a40588c-a3b2-4480-9ef4-ca8d6f874e84">